openfermioncirq/gates/four_qubit_gates.py

```diff
@@ -0,0 +1,165 @@
+                    cirq.CompositeGate,
+                    cirq.TextDiagrammable):
+    """Evolve under |0011><1100| + h.c. for some time."""
+
+    def __init__(self, *,  # Forces keyword args.
+                 half_turns: Optional[Union[cirq.Symbol, float]]=None,
+                 rads: Optional[float]=None,
+                 degs: Optional[float]=None,
+                 duration: Optional[float]=None) -> None:
+        """Initialize the gate.
+
+        At most one of half_turns, rads, degs, or duration may be specified.
+        If more are specified, the result is considered ambiguous and an
+        error is thrown. If no argument is given, the default value of one
+        half-turn is used.
+
+        Args:
+            half_turns: The exponent angle, in half-turns.
+            rads: The exponent angle, in radians.
+            degs: The exponent angle, in degrees.
+            duration: The exponent as a duration of time.
+        """
+
+        if len([1 for e in [half_turns, rads, degs, duration]
+                if e is not None]) > 1:
+            raise ValueError('Redundant exponent specification. '
+                             'Use ONE of half_turns, rads, degs, or duration.')
+
+        if duration is not None:
+            exponent = 2 * duration / numpy.pi
+        else:
+            exponent = cirq.value.chosen_angle_to_half_turns(
+                half_turns=half_turns,
+                rads=rads,
+                degs=degs)
+
+        super().__init__(exponent=exponent)
+
+    @property
+    def half_turns(self) -> Union[cirq.Symbol, float]:
+        return self._exponent
+
+    def _eigen_components(self):
+        minus_one_component = numpy.zeros((16, 16))
+        minus_one_component[3, 3] = minus_one_component[12, 12] = 0.5
+        minus_one_component[3, 12] = minus_one_component[12, 3] = 0.5
+
+        plus_one_component = numpy.zeros((16, 16))
+        plus_one_component[3, 3] = plus_one_component[12, 12] = 0.5
+        plus_one_component[3, 12] = plus_one_component[12, 3] = -0.5
+
+        return [
+            (-1, minus_one_component),
+            (1, plus_one_component),
+            (0, numpy.diag([1, 1, 1, 0, 1, 1, 1, 1,
+                            1, 1, 1, 1, 0, 1, 1, 1]))
+        ]
+
+    def _canonical_exponent_period(self) -> Optional[float]:
+        return 2
+
+    def _with_exponent(self,
+                       exponent: Union[cirq.Symbol, float]
+                       ) -> 'LocalPQRSGate':
+        return LocalPQRSGate(half_turns=exponent)
+
+    def default_decompose(self, qubits):
+        p, q, r, s = qubits
+
+        yield cirq.CNOT(r, s)
+        yield cirq.CNOT(q, p)
+        yield cirq.CNOT(q, r)
+
+        yield cirq.X(q) ** self.half_turns
+        yield cirq.Z(q) ** 0.125
+        yield cirq.CNOT(r, q)
+        yield cirq.Z(q) ** -0.125
+
+        yield cirq.CNOT(s, r)
+        yield cirq.CNOT(r, q)
+        yield cirq.CNOT(s, r)
+
+        yield cirq.Z(q) ** 0.125
```

Comment:
You have a pretty substantial chunk of gates repeated four times. Pick the redundant subset, assign it into a list, and yield the list instead of the gates individually.

Reply:
Neat! That's a lot shorter. I didn't realize I'd be able to yield a list of gates here (it even looks like arbitrarily nested lists are OK).

Reply:
You've just discovered cirq.OP_TREE

Reply:
I see, good to know why that worked ;)